### PR TITLE
Tighten condition when to check enum constructors

### LINKED ISSUE #11081: Incorrect "Illegal reference to type from enum case" in companion object
Not quite sure what's happening here, as I'm in the companion object, where I would expect stuff to behave like any other object.

## Minimized code

```Scala
enum Outer:
  case Foo
object Outer:
  trait Bar
  case class Baz(bar: Bar)
```

## Output

```scala
illegal reference to trait Bar from enum case
```

## Expectation
It compiles, or at least gives a better error message.

### PR DESCRIPTION
**Pull request by odersky**
Fixes #11081

### CODE REVIEW

File: tests/pos/i11081.scala [hunk 1]
```diff
@@ -0,0 +1,5 @@
+enum Outer:
+  case Foo
+object Outer:
+  trait Bar
+  case class Baz(bar: Bar)
```
Comment (bishabosha) [inline tests/pos/i11081.scala:5]: it still fails if instead there is a default argument to Baz's constructor that references Bar, I can add the extra check

**Review by bishabosha — CHANGES_REQUESTED** [review]

File: tests/pos/i11081.scala (hunk 1 above)
Comment (odersky) [reply tests/pos/i11081.scala]: Ah yes, I overlooked that part. Well spotted.

**Comment by bishabosha** [comment]
there should also be a change to `def isCase` on line 1223 to prevent the following:
```scala
enum Outer:
  case Foo(u: Unavailable) // ok
  case Bar(u: DefinitelyNotAvailable) // error
object Outer:
  case class Unavailable(i: Int)
  class DefinitelyNotAvailable
```

**Review by bishabosha — APPROVED** [review]
LGTM

**Pull request merged**